pkg/dockerregistry/server/tagservice.go

```diff
@@ -0,0 +1,213 @@
+				continue
+			}
+		}
+		tags = append(tags, tag)
+	}
+	return tags, nil
+}
+
+func (t tagService) Lookup(ctx context.Context, desc distribution.Descriptor) ([]string, error) {
+	tags := []string{}
+
+	imageStream, err := t.repo.getImageStream()
+	if err != nil {
+		context.GetLogger(ctx).Errorf("error retrieving ImageStream %s/%s: %v", t.repo.namespace, t.repo.name, err)
+		return tags, distribution.ErrRepositoryUnknown{Name: t.repo.Named().Name()}
+	}
+
+	managedImages := make(map[string]bool)
+
+	for tag, history := range imageStream.Status.Tags {
+		if len(history.Items) == 0 {
+			continue
+		}
+
+		dgst, err := digest.ParseDigest(history.Items[0].Image)
+		if err != nil {
+			context.GetLogger(ctx).Errorf("bad digest %s: %v", history.Items[0].Image, err)
+			continue
+		}
+
+		if dgst != desc.Digest {
+			continue
+		}
+
+		if !t.repo.pullthrough {
+			managed, found := managedImages[history.Items[0].Image]
+			if !found {
+				image, err := t.repo.getImage(dgst)
+				if err != nil {
+					context.GetLogger(ctx).Errorf("unable to get image %s/%s %s: %v", t.repo.namespace, t.repo.name, dgst.String(), err)
+					continue
+				}
+				managed = isImageManaged(image)
+				managedImages[history.Items[0].Image] = managed
+			}
+
+			if !managed {
+				continue
+			}
+		}
+		tags = append(tags, tag)
+	}
+
+	return tags, nil
+}
+
+func (t tagService) Tag(ctx context.Context, tag string, dgst distribution.Descriptor) error {
+	imageStream, err := t.repo.getImageStream()
+	if err != nil {
+		context.GetLogger(ctx).Errorf("error retrieving ImageStream %s/%s: %v", t.repo.namespace, t.repo.name, err)
+		return distribution.ErrRepositoryUnknown{Name: t.repo.Named().Name()}
+	}
+
+	image, err := t.repo.registryOSClient.Images().Get(dgst.Digest.String())
+	if err != nil {
+		context.GetLogger(ctx).Errorf("unable to get image: %s", dgst.Digest.String())
+		return err
+	}
+	image.SetResourceVersion("")
+
+	if !t.repo.pullthrough && !isImageManaged(image) {
+		return distribution.ErrRepositoryUnknown{Name: t.repo.Named().Name()}
+	}
+
+	ism := imageapi.ImageStreamMapping{
+		ObjectMeta: kapi.ObjectMeta{
+			Namespace: imageStream.Namespace,
+			Name:      imageStream.Name,
+		},
+		Tag:   tag,
+		Image: *image,
+	}
+
+	err = t.repo.registryOSClient.ImageStreamMappings(imageStream.Namespace).Create(&ism)
+	if quotautil.IsErrorQuotaExceeded(err) {
+		context.GetLogger(ctx).Errorf("denied creating ImageStreamMapping: %v", err)
+		return distribution.ErrAccessDenied
+	}
+
+	return err
+}
+
+func (t tagService) Untag(ctx context.Context, tag string) error {
+	imageStream, err := t.repo.getImageStream()
+	if err != nil {
+		context.GetLogger(ctx).Errorf("error retrieving ImageStream %s/%s: %v", t.repo.namespace, t.repo.name, err)
+		return distribution.ErrRepositoryUnknown{Name: t.repo.Named().Name()}
+	}
+
+	te := imageapi.LatestTaggedImage(imageStream, tag)
+	if te == nil {
+		return distribution.ErrTagUnknown{Tag: tag}
+	}
+
+	if !t.repo.pullthrough {
```

Comment:
Similar to the `Tag()` case. It can be invoked only from [`imageManifestHandler.DeleteImageManifest()`](https://github.com/miminar/distribution/blob/master/registry/handlers/images.go#L379). Not something client can call directly.